feat(CLI-alpha): support backfill for cron workflow. Part of #2706

shuangkun opened the pull request:
<!-- markdownlint-disable MD041 -- this is rendered within existing HTML, so allow starting without an H1 -->

Base #4212

Airflow is easy to backfill. Simplify the use of backfill on Argo.

Support：
```
Provide CLI like argo corn backfill <cronworkflow name> --name <backfill name> --start <startdate> --end <enddate>
```

Part of #2706

<!--

### Before you open your PR

- Run `make pre-commit -B` to fix codegen and lint problems (build will fail).
- [Signed-off your commits](https://github.com/apps/dco/) (otherwise the DCO check will fail).
- Used [a conventional commit message](https://www.conventionalcommits.org/en/v1.0.0/).

### When you open your PR

- PR title format should also conform to [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).
- "Fixes #" is in both the PR title (for release notes) and this description (to automatically link and close the issue).
- Create the PR as draft.
- Once builds are green, mark your PR "Ready for review".

When changes are requested, please address them and then dismiss the review to get it reviewed again.

-->

<!-- Does this PR fix an issue -->

### Motivation

<!-- TODO: Say why you made your changes. -->

### Modifications

<!-- TODO: Say what changes you made. -->
<!-- TODO: Attach screenshots if you changed the UI. -->

### Verification

<!-- TODO: Say how you tested your changes. -->

```
U-4YKHFNR6-2229 cron-backfill % ./argo cron backfill daily-job --start "Wed, 21 Oct 2024 15:28:00 GMT" --end "Wed, 28 Oct 2024 15:28:00 GMT" --argname data --parallel true
Created 28816 Backfill task for Cronworkflow daily-job
==================================================
Backfill Period :
Start Time : Wed, 21 Oct 2024 15:28:00 GMT
  End Time : Wed, 28 Oct 2024 15:28:00 GMT
Total Backfill Schedule: 7
==================================================
Backfill Workflows:
   NAMESPACE	 WORKFLOW:
1. argo 	 backfill-wf-jhmpr
tianshuangkun@U-4YKHFNR6-2229 cron-backfill % ./argo get backfill-wf-jhmpr
Name:                backfill-wf-jhmpr
Namespace:           argo
ServiceAccount:      unset (will run with the default ServiceAccount)
Status:              Running
Conditions:
 PodRunning          True
Created:             Tue Dec 17 14:44:02 +0800 (10 seconds ago)
Started:             Tue Dec 17 14:44:02 +0800 (10 seconds ago)
Duration:            10 seconds
Progress:            0/7

STEP                                                     TEMPLATE         PODNAME                                       DURATION  MESSAGE
 ● backfill-wf-jhmpr                                     main
 └─┬─● create-workflow(0:2024-10-22 02:00:00 +0000 GMT)  create-workflow  backfill-wf-jhmpr-create-workflow-3441413642  10s
   ├─● create-workflow(1:2024-10-23 02:00:00 +0000 GMT)  create-workflow  backfill-wf-jhmpr-create-workflow-2444440844  10s
   ├─● create-workflow(2:2024-10-24 02:00:00 +0000 GMT)  create-workflow  backfill-wf-jhmpr-create-workflow-2857101642  10s
   ├─● create-workflow(3:2024-10-25 02:00:00 +0000 GMT)  create-workflow  backfill-wf-jhmpr-create-workflow-2922680488  10s
   ├─● create-workflow(4:2024-10-26 02:00:00 +0000 GMT)  create-workflow  backfill-wf-jhmpr-create-workflow-4231395098  10s
   ├─● create-workflow(5:2024-10-27 02:00:00 +0000 GMT)  create-workflow  backfill-wf-jhmpr-create-workflow-2220838812  10s
   └─● create-workflow(6:2024-10-28 02:00:00 +0000 GMT)  create-workflow  backfill-wf-jhmpr-create-workflow-1221315226  10s
U-4YKHFNR6-2229 cron-backfill % argo list
NAME                             STATUS      AGE   DURATION   PRIORITY   MESSAGE
daily-job-backfill-28816-njt45   Running     11s   11s        0
daily-job-backfill-28816-shk9g   Running     11s   11s        0
daily-job-backfill-28816-vk79g   Running     11s   11s        0
backfill-wf-jhmpr                Running     13s   13s        0
daily-job-backfill-28816-k5wcm   Succeeded   11s   7s         0
daily-job-backfill-28816-q949s   Succeeded   11s   7s         0
daily-job-backfill-28816-xssbb   Succeeded   11s   7s         0
daily-job-backfill-28816-csfld   Succeeded   11s   6s         0
U-4YKHFNR6-2229 cron-backfill % argo list
NAME                             STATUS      AGE   DURATION   PRIORITY   MESSAGE
backfill-wf-jhmpr                Succeeded   34s   30s        0
daily-job-backfill-28816-vk79g   Succeeded   32s   22s        0
daily-job-backfill-28816-shk9g   Succeeded   32s   17s        0
daily-job-backfill-28816-njt45   Succeeded   32s   13s        0
U-4YKHFNR6-2229 cron-backfill % ./argo get backfill-wf-jhmpr
Name:                backfill-wf-jhmpr
Namespace:           argo
ServiceAccount:      unset (will run with the default ServiceAccount)
Status:              Succeeded
Conditions:
 PodRunning          False
 Completed           True
Created:             Tue Dec 17 14:44:02 +0800 (37 seconds ago)
Started:             Tue Dec 17 14:44:02 +0800 (37 seconds ago)
Finished:            Tue Dec 17 14:44:32 +0800 (7 seconds ago)
Duration:            30 seconds
Progress:            7/7
ResourcesDuration:   9s*(1 cpu),1m7s*(100Mi memory)

STEP                                                     TEMPLATE         PODNAME                                       DURATION  MESSAGE
 ✔ backfill-wf-jhmpr                                     main
 └─┬─✔ create-workflow(0:2024-10-22 02:00:00 +0000 GMT)  create-workflow  backfill-wf-jhmpr-create-workflow-3441413642  23s
   ├─✔ create-workflow(1:2024-10-23 02:00:00 +0000 GMT)  create-workflow  backfill-wf-jhmpr-create-workflow-2444440844  13s
   ├─✔ create-workflow(2:2024-10-24 02:00:00 +0000 GMT)  create-workflow  backfill-wf-jhmpr-create-workflow-2857101642  28s
   ├─✔ create-workflow(3:2024-10-25 02:00:00 +0000 GMT)  create-workflow  backfill-wf-jhmpr-create-workflow-2922680488  18s
   ├─✔ create-workflow(4:2024-10-26 02:00:00 +0000 GMT)  create-workflow  backfill-wf-jhmpr-create-workflow-4231395098  13s
   ├─✔ create-workflow(5:2024-10-27 02:00:00 +0000 GMT)  create-workflow  backfill-wf-jhmpr-create-workflow-2220838812  13s
   └─✔ create-workflow(6:2024-10-28 02:00:00 +0000 GMT)  create-workflow  backfill-wf-jhmpr-create-workflow-1221315226  13s
```

<!--
### Beyond this PR

Thank you for submitting this! Have you ever thought of becoming a Reviewer or Approver on the project?

Argo Workflows is seeking more community involvement and ultimately more [Reviewers and Approvers](https://github.com/argoproj/argoproj/blob/main/community/membership.md) to help keep it viable.
See [Sustainability Effort](https://github.com/argoproj/argo-workflows/blob/main/community/sustainability_effort.md) for more information.

-->

tczhao commented on test/e2e/fixtures/e2e_suite.go:
```diff
@@ -187,6 +188,17 @@ func (s *E2ESuite) DeleteResources() {
+			Namespace:         Namespace,
+			LabelRequirements: parse,
+		})
+		s.CheckError(err)
```
should we check backfill workflow actually exist, incase this returns empty list

shuangkun replied:
I referred to the previous implementation of cleaning/test label. If there is no error here, it doesn’t matter even if an empty list is returned. If the range of the subsequent loop is empty, delete will not be executed.

tczhao commented on docs/cli/argo_cron_backfill.md:
```diff
@@ -0,0 +1,59 @@
+      --format string    Date format for Schedule time value (default "Mon, 02 Jan 2006 15:04:05 MST")
+  -h, --help             help for backfill
+      --name string      Backfill name
+      --parallel         Enabled all backfile workflows run parallel
```
```suggestion
      --parallel         Enabled all backfill workflows run parallel
```

tczhao commented on cmd/argo/commands/cron/backfill.go:
```diff
@@ -0,0 +1,254 @@
+	return nil
+}
+
+var backfillWf = `{
```
```suggestion
const backfillWf = `{
```

tczhao commented on cmd/argo/commands/cron/backfill.go:
```diff
@@ -0,0 +1,254 @@
+	var scheList []string
+	wf := common.ConvertCronWorkflowToWorkflow(cronWF)
+	paramArg := `{{inputs.parameters.backfillscheduletime}}`
+	wf.GenerateName = cronWF.Name + "-backfill-" + strings.ToLower(cliOps.name) + "-"
```
we may want to truncate this in case this gets too long

shuangkun replied:
Add a function GenerateBackfillWorkflowPrefix

tczhao commented on docs/cli/argo_cron_backfill.md:
````diff
@@ -0,0 +1,59 @@
+### Options
+
+```
+      --argname string   Schedule time argument name for workflow (default "cronScheduleTime")
````
wondering if we really need this, it's only for parent workflow, right?

shuangkun replied:
Yes，This is useful when the sub-workflow needs to accept time parameters, such as pulling data based on time parameters.

tczhao commented on cmd/argo/commands/cron/backfill.go:
```diff
@@ -0,0 +1,254 @@
+`
+
+func CreateMonitorWf(ctx context.Context, wf, namespace, cronWFName string, scheTime []string, wfClient workflow.WorkflowServiceClient, cliOps backfillOpts) error {
+	const maxWfCount = 1000
```
need to expose this as an argument?

tczhao reviewed (commented):
I wonder if we should have a way to label this as something like a alpha feature,
so that we don't need to maintain backward compatibility incase there's a significant change later on

shuangkun commented:
> I wonder if we should have a way to label this as something like a alpha feature, so that we don't need to maintain backward compatibility incase there's a significant change later on

Yes, it would be better if there could be an alpha feature label. I have now added an alpha field in the pr title.

tczhao commented:
> > I wonder if we should have a way to label this as something like a alpha feature, so that we don't need to maintain backward compatibility incase there's a significant change later on
> 
> Yes, it would be better if there could be an alpha feature label. I have now added an alpha field in the pr title.

Perhaps let's add it to the CLI Short/Long description

shuangkun commented:
> > > I wonder if we should have a way to label this as something like a alpha feature, so that we don't need to maintain backward compatibility incase there's a significant change later on
> > 
> > 
> > Yes, it would be better if there could be an alpha feature label. I have now added an alpha field in the pr title.
> 
> Perhaps let's add it to the CLI Short/Long description

Added alpha to CLI short description

frosk1 commented:
Hey, when will this be merged and added to an official version with documentation and everything?

shuangkun commented:
> Hey, when will this be merged and added to an official version with documentation and everything?

It should be soon. Once it is merged, you can use the latest version for trial.
The official release should be in 3.7

shuangkun commented:
@tczhao Could you help merge it， thanks！
